### PR TITLE
feat: integrate data processor

### PR DESCRIPTION
**Pull request by jfsantos-ds**
Works on top of the base data processor and integrates it on the model structures and the examples.
When the processor is defined it is leveraged in the sample method to return synthetic records with the original format.

### CODE REVIEW

**Comment by jfsantos-ds** [comment]
Regular and TimeSeries processors integrated in the BaseModel train method.
Adult dragan example and Dragan model adapted with the new processors.

Todo:
- Integrate BaseModel train method on other regular synthesizers
- Develop the TimeSeries processor (issue #119 to be worked in a different PR)
- Integrate BaseModel train method on timeseries synthesizers
- Remove unused transformation scripts from codebase

**Review by fabclmnt — APPROVED** [review]
Approved. 🚀  

But please change the location of the RegularSynth enums.

File: src/ydata_synthetic/synthesizers/gan.py [hunk 1]
```diff
@@ -75,8 +88,34 @@ def model_name(self):
...
+        Args:
+            data (Union[DataFrame, array]): Raw data object.
+            num_cols (Optional[List[str]]): List of names of numerical columns.
+            cat_cols (Optional[List[str]]): List of names of categorical columns.
+            preprocess (bool): Determines if the preprocessor is to be run on the data or not (p.e. preprocessed data).
+        Returns:
+            processed_data (Union[DataFrame, array]): Processed data object (if applicable).
+        """
+        processed_data = data
+        if preprocess:
```
Comment (fabclmnt) [inline src/ydata_synthetic/synthesizers/gan.py:105]: Please perform the following change. 

```
if preprocess:
            if self.__MODEL__ in RegularModels.__members__:
                self.processor = RegularDataProcessor(num_cols, cat_cols)
            elif self.__MODEL__ in TimeSeriesModels.__members__:
                self.processor = TimeSeriesDataProcessor(num_cols, cat_cols)
            else:
                print(f'A DataProcessor is not available for the {self.__MODEL__}.')
            self.processor(num_cols, cat_cols).fit(data)
```

File: src/ydata_synthetic/synthesizers/gan.py [hunk 2]
```diff
@@ -75,8 +88,34 @@ def model_name(self):
...
+            if self.__MODEL__ in RegularModels.__members__:
+                self.processor = RegularDataProcessor
+            elif self.__MODEL__ in TimeSeriesModels.__members__:
+                self.processor = TimeSeriesDataProcessor
+            else:
+                print(f'A DataProcessor is not available for the {self.__MODEL__}.')
+        if self.processor and preprocess:
+            processed_data = self.processor(num_cols, cat_cols).fit_transform(data)
+
+        self.data_dim = processed_data.shape[1]
```
Comment (fabclmnt) [inline src/ydata_synthetic/synthesizers/gan.py:115]: Move this logic to the synthesizers. As this logic can be different for time-series synths.

**Review by fabclmnt — CHANGES_REQUESTED** [review]

File: examples/regular/adult_wgangp.py [hunk 3]
```diff
@@ -1,20 +1,26 @@
...
-noise_dim = 32
+noise_dim = 128
 dim = 128
-batch_size = 128
+batch_size = 500
 
 log_step = 100
 epochs = 300+1
-learning_rate = [5e-4, 3e-3]
+learning_rate = 1e-5
```
Comment (fabclmnt) [inline examples/regular/adult_wgangp.py:23]: Why have we changed this parameter to a single instead of a tuple as it was in the example? 

If I recall correctly this was meant to showcase that we can provide different learning rates for the Generator and Discriminator.

File: src/ydata_synthetic/synthesizers/regular/cgan/model.py [hunk 4]
```diff
@@ -65,23 +81,30 @@ def get_data_batch(self, train, batch_size, seed=0):
         stop_i = start_i + batch_size
         shuffle_seed = (batch_size * seed) // len(train)
         np.random.seed(shuffle_seed)
-        train_ix = np.random.choice(list(train.index), replace=False, size=len(train))  # wasteful to shuffle every time
+        train_ix = np.random.choice(train.shape[0], replace=False, size=len(train))  # wasteful to shuffle every time
```
Comment (fabclmnt) [inline src/ydata_synthetic/synthesizers/regular/cgan/model.py:84]: it shouldn't be from shape, this will create problems if the index is not ordered. 

I would recommend to revert to index as this was an issue spotted in the past.

File: src/ydata_synthetic/synthesizers/regular/cgan/model.py [hunk 5]
```diff
@@ -65,23 +81,30 @@ def get_data_batch(self, train, batch_size, seed=0):
...
+            train_arguments: GAN training arguments.
+            num_cols: List of columns of the data object to be handled as numerical
+            cat_cols: List of columns of the data object to be handled as categorical
+            preprocess: If True preprocess the data before using in train session
         """
-        iterations = int(abs(data.shape[0] / self.batch_size) + 1)
+        super().train(data, num_cols, cat_cols, preprocess)
+
+        processed_data = self.processor.transform(data)
+        self.data_dim = processed_data.shape[1] - len(self.processor.cat_pipeline.get_feature_names_out())
```
Comment (fabclmnt) [inline src/ydata_synthetic/synthesizers/regular/cgan/model.py:102]: From the data dim, why are we removing cat columns from the preprocessor dimension. 

I should remind that the class column shouldn't go through the preprocessing.

File: src/ydata_synthetic/synthesizers/regular/cgan/model.py [hunk 6]
```diff
@@ -124,10 +147,22 @@ def train(self, data: Union[DataFrame, array],
...
-                #Here is generating synthetic data
-                z = tf.random.normal((432, self.noise_dim))
-                label_z = tf.random.uniform((432,), minval=min(train_arguments.labels), maxval=max(train_arguments.labels)+1, dtype=tf.dtypes.int32)
-                gen_data = self.generator([z, label_z])
+                #Here is generating synthetic data from an arbitrary condition
+                gen_data = self.sample(label[0], 1000)
+
+    def sample(self, cond_array, n_samples):
+        """Produce n_samples by conditioning the generator with cond_array."""
+        assert cond_array.shape[0] == self.num_classes, \
```
Comment (fabclmnt) [inline src/ydata_synthetic/synthesizers/regular/cgan/model.py:155]: The conditional array should be a single array, with the len of the sample to be generated. 

cond_array rows = n_samples.

File: src/ydata_synthetic/synthesizers/regular/cgan/model.py [hunk 7]
```diff
@@ -136,9 +171,9 @@ def __init__(self, batch_size, num_classes):
 
     def build_model(self, input_shape, dim, data_dim):
         noise = Input(shape=input_shape, batch_size=self.batch_size)
-        label = Input(shape=(1,), batch_size=self.batch_size, dtype='int32')
-        label_embedding = Flatten()(Embedding(self.num_classes, 1)(label))
-        input = multiply([noise, label_embedding])
+        label = Input(shape=(self.num_classes,), batch_size=self.batch_size, dtype='int32')
+        label_embedding = Embedding(2, 1)(label)
```
Comment (fabclmnt) [inline src/ydata_synthetic/synthesizers/regular/cgan/model.py:175]: Not understanding the need to remove the flatten here. 

No changes should be done at the level of the generator. The processing was already done before, so nothing should change with the new processor. 

Please rever the Generator changes.

**Review by fabclmnt — CHANGES_REQUESTED** [review]
Please revise the conditional GAN flow. 

Is not correct after the processor integration.

File: src/ydata_synthetic/synthesizers/regular/cgan/model.py (hunk 4 above)
Comment (jfsantos-ds) [reply src/ydata_synthetic/synthesizers/regular/cgan/model.py:84]: Now I remember the reason for this change, after the preprocessing the object train is no longer a pd.DataFrame but a np.array.
I believe this change has no negative impact in get_data_batch functionality since we are using numpy's positional indexing instead of panda's loc index. For the purpose of shuffling it should not be affected by the index order. Let me know if I am missing something

File: src/ydata_synthetic/synthesizers/regular/cgan/model.py [hunk 8]
```diff
@@ -85,7 +83,7 @@ def get_data_batch(self, train, batch_size, seed=0):
         train_ix = list(train_ix) + list(train_ix)  # duplicate to cover ranges past the end of the set
         return train[train_ix[start_i: stop_i]]
 
-    def train(self, data: Union[DataFrame, array], label:str, train_arguments: TrainParameters, num_cols: List[str],
+    def train(self, data: Union[DataFrame, array], label_col: str, train_arguments: TrainParameters, num_cols: List[str],
               cat_cols: List[str], preprocess: bool = True):
```
Comment (jfsantos-ds) [inline src/ydata_synthetic/synthesizers/regular/cgan/model.py:87]: Assumption 1: All data is passed in the data argument, the condition columns are included in this object
Assumption 2: A single column from the original data is expected to be identified by the label_col str. Not a priority but eventually this should be easily extendable for a list of conditions p.e.

File: src/ydata_synthetic/synthesizers/regular/cgan/model.py [hunk 9]
```diff
@@ -102,6 +100,9 @@ def train(self, data: Union[DataFrame, array], label:str, train_arguments: Train
         self.data_dim = processed_data.shape[1] - len(self.processor.cat_pipeline.get_feature_names_out())
         self.define_gan()
 
+        mask = zeros(processed_data.shape[1], bool)  # Identify column indexes that carry the processed label feature
+        mask[[i for i, col in enumerate(list(self.processor._num_pipeline.get_feature_names_out()) + list(self.processor._cat_pipeline.get_feature_names_out())) if ''.join(col.split('_')[:-1]) == label_col]] = True
+
```
Comment (jfsantos-ds) [inline src/ydata_synthetic/synthesizers/regular/cgan/model.py:105]: After the whole data object is processed, this mask is created flagging the processed output columns that originate from label_col.

File: src/ydata_synthetic/synthesizers/regular/cgan/model.py [hunk 10]
```diff
@@ -115,15 +116,14 @@ def train(self, data: Union[DataFrame, array], label:str, train_arguments: Train
                 #  Train Discriminator
                 # ---------------------
                 batch_x = self.get_data_batch(processed_data, self.batch_size)
-                label = batch_x[:, self.processor._num_col_idx_: self.processor._cat_col_idx_]
-                data_cols = list(range(self.processor._num_col_idx_))  # All data without the label columns
+                batch_x, label = batch_x[:, ~mask], batch_x[:, mask]
```
Comment (jfsantos-ds) [inline src/ydata_synthetic/synthesizers/regular/cgan/model.py:119]: Split the batch on records without condition and the conditions itself.

File: src/ydata_synthetic/synthesizers/regular/cgan/model.py [hunk 11]
```diff
@@ -172,8 +172,8 @@ def __init__(self, batch_size, num_classes):
     def build_model(self, input_shape, dim, data_dim):
         noise = Input(shape=input_shape, batch_size=self.batch_size)
         label = Input(shape=(self.num_classes,), batch_size=self.batch_size, dtype='int32')
-        label_embedding = Embedding(2, 1)(label)
-        input = Flatten()(multiply([noise, label_embedding]))
+        label_dense = Dense(1)(label)
```
Comment (jfsantos-ds) [inline src/ydata_synthetic/synthesizers/regular/cgan/model.py:175]: Changed Embedding layer to Dense layer with same dimension, numerically outputs are expected to be the same. We should be able to process both numeric and categoric conditions this way. [Computationally should be more expensive.](https://stackoverflow.com/questions/47868265/what-is-the-difference-between-an-embedding-layer-and-a-dense-layer)

File: src/ydata_synthetic/synthesizers/regular/cgan/model.py [hunk 12]
```diff
@@ -172,8 +172,8 @@ def __init__(self, batch_size, num_classes):
...
+        input = multiply([noise, label_dense])
```
Comment (jfsantos-ds) [inline src/ydata_synthetic/synthesizers/regular/cgan/model.py:176]: Flatten was no longer required since no extra dimensions are added to the processed label.

**Comment by fabclmnt** [comment]
💄 Suggested improvement:
- Validate wether the label column is of a valid type.

**Review by fabclmnt — APPROVED** [review]
Only a minor improvement. 

🚀
